cmd/snap-update-ns/utils.go

```diff
@@ -273,27 +258,39 @@ func (sec *Secure) MkDir(dirFd int, dirName string, name string, perm os.FileMod
 			sysClose(newFd)
 			return -1, fmt.Errorf("cannot chown directory %q to %d.%d: %v", filepath.Join(dirName, name), uid, gid, err)
 		}
+		// As soon as we find a place that is safe to write we can switch off
+		// the restricted mode (and thus any subsequent checks). This is
+		// because we only allow "writing" to read-only filesystems where
+		// writes fail with EROFS or to a tmpfs that snapd has privately
+		// mounted inside the per-snap mount namespace. As soon as we start
+		// walking over such tmpfs any subsequent children are either read-
+		// only bind mounts from $SNAP, other tmpfs'es  (e.g. one explicitly
+		// constructed for a layout) or writable places that are bind-mounted
+		// from $SNAP_DATA or similar.
+		rs.LiftRestrictions()
```

Comment:
Wondering if the caller should do it and instead `MkDir()` should return if the directory was made or not. Modifying this inside in non-obvious way makes me feel a bit uncomfortable.

Reply:
I explicitly didn't want to do this. In an earlier iteration of this code I was passing and returning the restricted flag around and it was simply more confusing and convoluted to handle. Specifically this is an internal implementation detail of the restriction system and in fact, mkdir is the only place where the restriction can be safely lifted so it felt like pushing the responsibility to the caller.